### PR TITLE
Transformation

### PR DESCRIPTION
**Pull request by Addehman**
Created the script for Transforming the player character from werewolf to human.

Transform by holding "R"-key for 1.2 seconds which completes the transformation into human form. After 5 seconds it will return to the werewolf form again.

## Visualization
transformation-show_00

### CODE REVIEW

File: Assets/Scripts/Transformation.cs [hunk 1]
```diff
@@ -0,0 +1,86 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class Transformation : MonoBehaviour
+{
+    //[SerializeField] private bool isTransformed = false;
```
Comment (MrJackey) [inline Assets/Scripts/Transformation.cs]: Is this comment required to exist? Do we need it?

File: Assets/Scripts/Transformation.cs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+    private void Start() {
+		mySpriteRend = gameObject.GetComponent<SpriteRenderer>();
+    }
+
+
+    private void Update() {
+		if (Input.GetKeyDown(KeyCode.R) && transStates == TransformationStates.Wolf) {
+            StartCoroutine(CoTransforming());
+			Debug.Log("Transformation process Charging...");
```
Comment (MrJackey) [inline Assets/Scripts/Transformation.cs]: Do you think debug logging should be kept?

**Review by MrJackey — CHANGES_REQUESTED** [review]
Looking good, just a lot of debug which maybe should be removed

File: Assets/Scripts/Transformation.cs (hunk 1 above)
Comment (smeas) [inline Assets/Scripts/Transformation.cs]: Brace should be on the same line.

File: Assets/Scripts/Transformation.cs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+			  mySpriteRend.color = humanColor;
+        }
+        else if (transStates == TransformationStates.Wolf) {
+			mySpriteRend.color = wolfColor;
+        }
+    }
+
+
+	private IEnumerator CoTransforming() {
+
```
Comment (smeas) [inline Assets/Scripts/Transformation.cs]: Remove empty line at start of method.

File: Assets/Scripts/Transformation.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+			TransformToHuman();
+			StartCoroutine(HumanFormDuration());
+		}
+		else if (transStates == TransformationStates.Human) {
+			TransformToWolf();
+		}
+	}
+
+
+	private IEnumerator HumanFormDuration() {
```
Comment (smeas) [inline Assets/Scripts/Transformation.cs]: Use 'Co' prefix for coroutine?

**Review by smeas — CHANGES_REQUESTED** [review]
Just some formatting/style issues.

There is mixed whitespace (spaces/tabs) in the whole file.

File: Assets/Scripts/Transformation.cs (hunk 1 above)
Comment (Addehman) [reply Assets/Scripts/Transformation.cs]: Ah nope!

Comment (Addehman) [reply Assets/Scripts/Transformation.cs]: ah ofc!

File: Assets/Scripts/Transformation.cs (hunk 2 above)
Comment (Addehman) [reply Assets/Scripts/Transformation.cs]: consider them gone!

File: Assets/Scripts/Transformation.cs (hunk 3 above)
Comment (Addehman) [reply Assets/Scripts/Transformation.cs]: done!

File: Assets/Scripts/Transformation.cs (hunk 4 above)
Comment (Addehman) [reply Assets/Scripts/Transformation.cs]: ah! right! thanks!

**Review by Addehman — COMMENTED** [review]
Applied the changes! thanks for the reviews!

File: Assets/Scripts/Transformation.cs [hunk 5]
```diff
@@ -25,44 +23,44 @@ public class Transformation : MonoBehaviour
...
-				Debug.Log("Transformation process interrupted");
+			if (Input.GetKeyUp(KeyCode.R) && transStates == TransformationStates.Transforming) {
+				transStates = TransformationStates.Wolf;
 				yield break;
 			}
 			yield return null;
 		}
 
-    	if (transStates == TransformationStates.Wolf) {
+    	if (transStates == TransformationStates.Transforming) {
```
Comment (MrJackey) [inline Assets/Scripts/Transformation.cs]: Are you sure this should be .Transforming? Shouldn't this remain as .Wolf?

File: Assets/Scripts/Transformation.cs [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+			mySpriteRend.color = transformingColor;
+		}
+		else if (transStates == TransformationStates.Wolf) {
+			mySpriteRend.color = wolfColor;
+		}
+	}
+
+
+	private IEnumerator CoTransforming() {
+		transStates = TransformationStates.Transforming;
```
Comment (MrJackey) [inline Assets/Scripts/Transformation.cs]: What about having this as a boolean instead? Would be more generic if you could see if transforming and keep it from a middle. Might depend on if the human to wolf transformation should be instant.

**Review by MrJackey — CHANGES_REQUESTED** [review]
Just one question, otherwise looks good

File: Assets/Scripts/Transformation.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+			mySpriteRend.color = transformingColor;
+
+			for (transTimer = transDuration; transTimer > 0 ; transTimer -= Time.deltaTime) {
+				if (Input.GetKeyUp(KeyCode.R)) {
+					mySpriteRend.color = humanColor;
+					yield break;
+				}
+				yield return null;
+			}
+		}
```
Comment (smeas) [inline Assets/Scripts/Transformation.cs]: Duplicate code in if statement blocks. The only difference is what color you set so you only need to check when you set the color.
```suggestion
		mySpriteRend.color = transformingColor;

		for (transTimer = transDuration; transTimer > 0 ; transTimer -= Time.deltaTime) {
			if (Input.GetKeyUp(KeyCode.R)) {
				if (transStates == TransformationStates.Wolf)
					mySpriteRend.color = wolfColor;
				else if (transStates == TransformationStates.Human)
					mySpriteRend.color = humanColor;

				yield break;
			}
			yield return null;
		}
```

File: Assets/Scripts/Transformation.cs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+		if (transStates == TransformationStates.Wolf) {
+			TransformToHuman();
+			StartCoroutine(CoHumanFormDuration());
+		}
+
```
Comment (smeas) [inline Assets/Scripts/Transformation.cs:51]: Remove empty line in the middle of if-else chain.

**Review by smeas — CHANGES_REQUESTED** [review]
BUG: If you transform to wolf>human>wolf>human you get a shorter human form duration.
The problem is that the `CoHumanFormDuration` from the previous transformation is still running .

**Review by smeas — APPROVED** [review]

**Review by MrJackey — APPROVED** [review]
+1

**Review by smeas — APPROVED** [review]

**Pull request merged**